specs-derive/src/impl_saveload.rs

```diff
@@ -301,85 +369,137 @@ fn saveload_tuple_struct(
-                let names_3 = get_names();
+                let saveload_fields = convert_fields_to_metadata(&fields.named);
+
+                let field_ser = saveload_fields.iter().map(|field_meta| {
+                    let field_ident = &field_meta.field.ident;
+
+                    if field_meta.skip_field {
+                        quote!{ #field_ident: #field_ident.clone() }
+                    } else {
+                        quote!{ #field_ident: ConvertSaveload::convert_into(#field_ident, &mut ids)? }
+                    }
+                });
+
+                let names = fields
+                    .named
+                    .iter()
+                    .map(|f| f.ident.clone())
+                    .collect::<Vec<_>>();
 
                 big_match_ser = quote! {
                     #big_match_ser
-                    #name::#ident { #( ref #names ),* } => #saveload_name::#ident { #( #names_3: ConvertSaveload::convert_into(#names_2, ids)? ),* },
+                    #name::#ident { #( ref #names ),* } => #saveload_name::#ident { #(#field_ser),* },
                 };
 
-                let names = get_names();
-                let names_2 = get_names();
-                let names_3 = get_names();
+                let field_de = saveload_fields.iter().map(|field_meta| {
+                    let field_ident = &field_meta.field.ident;
+
+                    if field_meta.skip_field {
+                        quote!{ #field_ident: #field_ident }
+                    } else {
+                        quote!{ #field_ident: ConvertSaveload::convert_from(#field_ident, &mut ids)? }
+                    }
+                })
+                .collect::<Vec<_>>();
 
                 big_match_de = quote! {
                     #big_match_de
-                    #saveload_name::#ident { #( #names ),* } => #name::#ident { #( #names_3: ConvertSaveload::convert_from(#names_2, &mut ids)? ),* },
+                    #saveload_name::#ident { #( #names ),* } => #name::#ident { #(#field_de),* },
                 };
             }
             Fields::Unnamed(fields) => {
-                let field_ids: Vec<_> = fields
-                    .unnamed
+                let saveload_fields = convert_fields_to_metadata(&fields.unnamed);
+
+                let make_field_ident =
+                    |i: usize| Ident::new(&format!("field{}", i), data.enum_token.span);
+
+                let field_ser_ids = saveload_fields
                     .iter()
-                    .cloned()
                     .enumerate()
-                    .map(|(i, _)| Some(Ident::new(&format!("field{}", i), data.enum_token.span)))
-                    .collect();
-                let field_ids_2 = field_ids.clone();
-                let tmp = field_ids.clone();
+                    .map(|(i, _)| make_field_ident(i))
+                    .collect::<Vec<_>>();
+
+                let field_ser = saveload_fields
+                    .iter()
+                    .enumerate()
```

Comment:
can save on an `enumerate()` and `make_field_ident` by zipping:

```rust
saveload_fields
    .iter()
    .zip(field_ser_ids.iter())
    .map(|(field_ident, field_meta)| {
        if field_meta.skip_field {
            quote! { #field_ident.clone() }
        } else {
            quote! { ConvertSaveload::convert_into(#field_ident, &mut ids)? }
        }
    })
```

same for deserialization